CHANGELOG.d/feature_derive-foldable.md

```diff
@@ -0,0 +1,61 @@
+* Derive Foldable instances
+
+  The compiler can now automatically derive a `Foldable` instance for types,
+  so long as
+  1. such types do not include a function as one of their subterms
+  2. at least one of the constructor(s) in a type can pass an `a` value
+     into the fold's `f` function.
+
+  ```purs
+  data M f a
+    = M0
+    | M1 a (Array a)
+    | M2 Int
+    | M3 (f a)
+    -- Note: fold's order depends on label declaration order,
+    -- not sorted labels' order.
+    | M4 { a :: a, x :: Array a, fa :: f a
+         , ignore :: Int, no :: Array Int, nope :: f Int
+         -- nested labels will be recursed
+         , nested :: { anotherOne :: a }
+         }
+
+  derive instance Foldable f => Foldable (M f)
+  {-
+  -- generates something like this:
+  instance Foldable f => Foldable (M f) where
+    foldl f z = case _ of
+      M0 -> z
+      M1 a arr -> foldl f (f z a) arr
+      M2 _ -> z
+      M3 fa -> foldl f z fa
+      M4 rec ->
+        foldl f (foldl f (foldl f (f z rec.a) rec.x) rec.fa) rec.nested.anotherOne
+    foldr f z = case _ of
+      M0 -> z
+      M1 a arr -> f a (foldr f z arr)
+      M2 _ -> z
+      M3 fa -> foldr f z fa
+      M4 rec ->
+        f (foldr f (foldr f (f z rec.nested.anotherOne) rec.fa) rec.x) rec.a
+    foldMap f = case _ of
+      M0 -> mempty
+      M1 a arr -> f a <> foldMap f arr
+      M2 _ -> mempty
+      M3 fa -> foldMap f fa
+      M4 rec -> f rec.a <> foldMap f rec.x <> foldMap f rec.fa <> foldMap rec.nested.anotherOne
+  -}
+
+  data Nope1 a = Nope1 (Int -> a)
+  data Nope2 a = Nope2 (a -> Int)
+  data Nope3 a = Nope3
```

Comment:
Not sure if this reflects the current state of the code, but my position is that `Nope3` (a.k.a. `Proxy`) should allow `Functor` (it already does), `Foldable`, and `Traversable` to be derived.